feat: add tab bookmark count badge indicator

qixing-jk opened the pull request:
This PR introduces a new feature that displays a badge on the icon of the browser extension to show the number of times the current page's URL has been added as a bookmark.
 
## Main functions
**Bookmark quantity indicator:** The expansion icon now displays a symbol that contains the number of bookmarks in the active tab. If the current URL has been accurately bookmarked, the symbol is green; otherwise, it is red.
Real-time update: The count of the indicator will be automatically updated when bookmarks are added or removed for the current page.
**Configurable cache:** To optimize performance and reduce API calls, the indicator count is cached. The expiration time of the cache can be configured in the extension's options. We adopted the stale-while-revalidate (SWR) strategy to ensure the final consistency of the indicator count.
**Settings switch:** This indicator feature can be enabled or disabled in the extension's options page.
Cache management: A new menu item has been added to the right-click menu of the extension icon, allowing users to clear the cache for the current page or all pages.
**Improved URL verification:** More strict verification has been implemented, allowing bookmarks to be added only for URLs with the http and https protocols.


## Related issue
https://github.com/karakeep-app/karakeep/issues/486


## Results show
<img width="67" height="64" alt="image" src="https://github.com/user-attachments/assets/9090b2ac-a9cb-4ff0-83fc-c01884122973" />
<img width="64" height="55" alt="image" src="https://github.com/user-attachments/assets/7f0ca3c1-2da5-40a2-933a-623f0074f06f" />

You can now easily check if the page you're viewing has already been hoarded and see the total count of similar URLs.
Badge Numbers and Colors: Their Meanings
Bookmark Count Badge System

**Smart Color Coding:**
🟢 Green: Exact URL match found
🔴 Red: No bookmarks found
🔢 Numbers: 0 = No bookmarks, 1+ = Count of bookmarks for current page

**Real-time Updates**  
- The Badge updates instantly after adding a bookmark, no need to refresh the page manually.  
- When a bookmark is removed, the Badge status synchronizes immediately as well.

MohamedBassem commented on apps/browser-extension/src/background/background.ts:
```diff
@@ -1,124 +1,262 @@
+const OPEN_KARAKEEP_ID = "open-karakeep";
+const ADD_LINK_TO_KARAKEEP_ID = "add-link";
+
+// Initialize the cache system, which creates a timer for periodic cache cleanup.
+initializeCache();
```
Is the alarm really needed? Can we just store the last time the purge was done in the in memory cache and just trigger a purge async if an hour has passed in the `onUpdated` / `onActivated` callbacks?

MohamedBassem commented on apps/browser-extension/src/background/background.ts:
```diff
@@ -1,124 +1,262 @@
+
+// Listen for timer events and execute cache cleanup tasks when triggered.
+chrome.alarms.onAlarm.addListener((alarm) => {
+  if (alarm.name === "badgeCachePurgeAlarm") {
```
PURGE_ALARM_NAME?

MohamedBassem commented on apps/browser-extension/src/background/background.ts:
```diff
@@ -1,124 +1,262 @@
+    if (cachedInfo) {
+      const { count: cachedBadgeCount, isExisted: cachedIsExisted } =
+        cachedInfo;
+      await setBadge(cachedBadgeCount, cachedIsExisted, tabId);
```
Setting a count text is a bit confusing, can we just set a boolean indicator of whether the page existed or not? I don't think setting a count adds a value?

qixing-jk replied:
<img width="716" height="956" alt="image" src="https://github.com/user-attachments/assets/c773ab78-c207-416f-a726-080debb84f1f" />
<img width="690" height="776" alt="image" src="https://github.com/user-attachments/assets/3da51cd6-53b5-4f86-9b20-f264e9ccc363" />
This is the effect I want to achieve. The color indicates whether the links are exactly the same, and the number represents the total number of matching items. 
If possible, as in this case, present specific results

MohamedBassem commented on apps/browser-extension/src/background/background.ts:
```diff
@@ -1,124 +1,262 @@
+  if (msg && msg.type) {
+    if (
+      msg.currentTab &&
+      (msg.type === MessageType.BOOKMARK_CREATED_REFRESH_BADGE ||
+        msg.type === MessageType.BOOKMARK_DELETED_REFRESH_BADGE)
```
If we're handling them the same, do we need two different message types? Why not just `BOOKMARK_REFRESH_BADGE`?

MohamedBassem commented on apps/browser-extension/src/OptionsPage.tsx:
```diff
@@ -55,6 +60,24 @@ export default function OptionsPage() {
       <hr />
+      <div className="flex flex-col gap-2">
+        <span className="mb-1">Show count badge</span>
+        <Button
```
Can we use a toggle instead of a button?

MohamedBassem commented:
@qixing-jk the coderabbit stuff are experimental, feel free to ignore them if they don’t make sense. This it the first PR it runs on :)

qixing-jk commented:
> @qixing-jk the coderabbit stuff are experimental, feel free to ignore them if they don’t make sense. This it the first PR it runs on :)

Some of the suggestions are indeed meaningful, but I have just finished formulating my thoughts. Now I will start addressing the issues raised by these suggestions.

qixing-jk commented:
Sorry, after multiple rounds of SWR caching attempts, I found that the iterative process was completed only to find out that tanstack already has a ready-made swr mechanism, I should have looked more into the project package dependencies and package-specific features.

MohamedBassem commented:
Sorry it took too long to merge this PR!
